### PR TITLE
[Bug fix] tril/triu return NaN for non-finite inputs

### LINKED ISSUE #52038: [ttnn]: tril/triu are composed as a mask multiply, so an inf in the masked region becomes NaN instead of 0 (fp32)
### Component / Area

ttnn

### Issue Type (optional)

Bad Outputs

### Observed

Measured on ttsim, Tenstorrent's own simulator, in **float32**. I do not have Tenstorrent silicon; a maintainer pointed me at ttsim in #51845 and it runs the real kernel.

Scope, up front, because it is narrower than the composition alone would suggest: **this reproduces on the float32 path and not on the bfloat16 one.** On a 32x32 tensor filled with `-inf`, fp32 turns 496 of the 496 masked cells into NaN; bf16 turns 0 of 496 into NaN. I cannot tell whether the bf16 behaviour is the device's or the simulator's, so I am only claiming fp32.

`tril` selects the lower triangular region and **zeroes** the rest. It is composed as a multiply against a 0/1 mask (`unary_composite_op.cpp:308-318`):

```cpp
Tensor tril(const Tensor& input_a, std::int32_t diag, const std::optional<MemoryConfig>& output_mem_config) {
    Tensor index_l = ttnn::index_tril<::bfloat16>(
        input_a.logical_shape(), input_a.padded_shape(), diag,
        DataType::BFLOAT16, Layout::TILE, input_a.device(),
        output_mem_config.value_or(input_a.memory_config()));
    return ttnn::multiply(input_a, index_l, std::nullopt, output_mem_config);
}
```

Zeroing and multiplying by zero are not the same operation. In IEEE-754, `0 * ±inf` is `NaN`, and `0 * NaN` is `NaN`. So every masked-out element that happens to hold an infinity or a NaN comes out as **NaN** rather than `0`.

`triu` immediately below has the same shape with `index_triu`.

**Where this actually bites.** Causal attention masks are commonly built with `-inf` in the positions to be suppressed, and tril/triu are the natural way to build or apply them. Measured on a 32x32 float32 tensor filled with `-inf`:

| fill | cells the mask must zero | come out 0 | **come out NaN** |
|---|---|---|---|
| **-inf** | 496 | 0 | **496** |
| **+inf** | 496 | 0 | **496** |
| -1e30 (control) | 496 | **496** | 0 |
| -1.0 (control) | 496 | **496** | 0 |

**Every masked cell, not some of them.** The two controls are the point: with a large finite fill the mask works perfectly, so this is the infinity specifically and not the masking in general. The retained triangle is untouched in all four cases.

The rate is 100%, so the size does not matter — a 128x128 causal mask loses all 8,128 masked positions the same way. The NaNs then propagate through the softmax, so a whole row is lost rather than one element being slightly off.

`torch.tril` does not behave this way: it writes zeros into the masked region rather than scaling it, so on the same input the registered golden and the kernel disagree.

**On test coverage.** The sweep at `tests/sweep_framework/sweeps/eltwise/unary/tril/tril.py` generates inputs with `torch_random(low=-100, high=100)`, so it never produces an infinity or a NaN. That is why this has not surfaced: the failing region sits outside what the sweep samples, rather than the sweep accepting it.

### Expected

The masked region should be exactly `0` for every input, including `±inf` and `NaN`, matching `torch.tril` and `torch.triu`.

The invariant, checkable without any golden table:

```
tril(x)[i][j] == 0            for every j > i + diag
tril(x)[i][j] == x[i][j]      otherwise
```

A select is the natural composition and it already exists in ttnn: `ttnn::where(mask, input, 0.0f)` keeps the kept region bit-exact and writes a true zero into the rest, with no arithmetic on the masked values at all. That also removes the multiply's rounding on the kept side, which is a small bonus rather than the point.

If the multiply is preferred for performance reasons, the masked region still needs a guard so non-finite values do not survive it.

### 1. Steps (exact commands)

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

N = 32
x = torch.full((N, N), float("-inf"), dtype=torch.float32)   # fp32: see scope note
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
out = ttnn.to_torch(ttnn.tril(t))

upper = [out[i][j].item() for i in range(N) for j in range(N) if j > i]
print("masked cells:", len(upper))                                  # 496
print("NaN         :", sum(1 for v in upper if v != v))             # 496  -- expected 0
print("zero        :", sum(1 for v in upper if v == 0.0))           # 0    -- expected 496
print("torch NaNs  :", torch.tril(x).isnan().sum().item())          # 0

# control: same shape, large finite fill -> the mask works
y = torch.full((N, N), -1e30, dtype=torch.float32)
ty = ttnn.from_torch(y, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
oy = ttnn.to_torch(ttnn.tril(ty))
print("control NaN :", sum(1 for i in range(N) for j in range(N)
                           if j > i and oy[i][j].item() != oy[i][j].item()))   # 0
```

Minimal single-element version:

```python
x = torch.tensor([[1.0, float("inf")], [1.0, 1.0]], dtype=torch.float32)
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.tril(t)))   # [[1, nan], [1, 1]]  -- expected [[1, 0], [1, 1]]
print(torch.tril(x))                 # [[1, 0], [1, 1]]
```

Swapping `torch.float32`/`ttnn.float32` for bfloat16 in the snippet above gives 0 NaN and 496 zeros — the bf16 path does not reproduce, which is why the claim is scoped to fp32.

### 2. Input data / link or description

No data file. Inputs are constructed in the snippet: a 32x32 float32 tensor filled with `-inf`, plus a large-finite control. The 128x128 causal-mask case behaves identically, since the failure rate is 100%.

### 3. Frequency

Deterministic — 100% for any non-finite value landing in the masked region. It is a property of the composition, not a race or a flake.

### 1. Software Versions

Deterministic — 100% for any non-finite value landing in the masked region. It is a property of the composition, not a race or a flake.

### 2. Hardware Details

**No Tenstorrent silicon.** Everything above is measured on **ttsim** running the real kernel, in a stock `tt-metalium` release container, float32. The bfloat16 path does not reproduce there and I make no claim about it.

What that does and does not affect:

- The composition (`ttnn::multiply` against a 0/1 mask) is read from `unary_composite_op.cpp:308-318` in `main`. Not device-dependent.
- `0 * ±inf = NaN` and `0 * NaN = NaN` are IEEE-754. Not device-dependent.
- The 8,128 figure is a direct count of the masked positions in a 128x128 causal layout, not a simulation.
- The question of whether the SFPU multiply follows IEEE or flushes is already answered on the simulator, and the answer differs by dtype: **fp32 follows IEEE (496/496 NaN), bf16 does not (0/496)**. That split is the reason the claim is scoped to fp32.
- What is left to confirm on silicon is only whether the simulator's fp32 multiply matches the device's. If the device flushes there too, this does not reproduce and the issue can be closed — that single check settles it.

### Is this a regression?

No

### Regression Details

_No response_

### Logs & Diagnostics

No logs — there is no crash. The op returns NaN cleanly.

Same structural family, if it helps to batch the review:
- #51861 — `ttnn.hypot` returns `inf` because it computes `sqrt(a**2 + b**2)` literally. A composed op whose intermediate leaves the representable range where the final result would not.

### Priority

P2

### Impact

Affects `ttnn.tril` and `ttnn.triu` for any input containing `±inf` or `NaN` in the region being masked. The practical exposure is causal attention masking, which is where these ops are most used and where `-inf` padding is the convention — so the failing case sits close to the intended use rather than at an edge.

Because the output is `NaN` rather than a slightly wrong number, it propagates: one masked element poisons the softmax row that consumes it.

The fix is a re-composition using `ttnn::where`, which already exists, rather than a new kernel.

---

Written with AI assistance (Claude Opus 5 via Claude Code). I read the source, verified the IEEE behaviour and the mask counts myself, and I am responsible for the numbers and the claim as submitted.

### PR DESCRIPTION
**Pull request by EazyHood**
### Ticket

Fixes #52038.

### What's wrong

With a `float32` input, `ttnn.tril` and `ttnn.triu` return `NaN` in the masked-out triangle
whenever the input is not finite. The masked-out region is defined as zero, and the golden
functions registered for these ops are literally `torch.tril` and `torch.triu`, which return
zero there.

```
input = full([5, 5], -inf), triu(diagonal=1)

golden (torch.triu)              ttnn.triu today
  0  -inf -inf -inf -inf          nan  -inf -inf -inf -inf
  0    0  -inf -inf -inf          nan   nan -inf -inf -inf
  0    0    0  -inf -inf          nan   nan  nan -inf -inf
  0    0    0    0  -inf          nan   nan  nan  nan -inf
  0    0    0    0    0           nan   nan  nan  nan  nan
```

15 NaN where the golden has none.

### Cause

Both are composed as a multiply against a 0/1 index mask
(`unary_composite_op.cpp`):

```cpp
Tensor index_u = ttnn::index_triu<::bfloat16>(...);
return ttnn::multiply(input_a, index_u, std::nullopt, output_mem_config);
```

`index_trilu` fills that mask with a boolean cast — `int32_t value = (x >= (y + diag))` — so the
masked-out entries are exactly `0.0`. Multiplication is equivalent to masking only for finite
operands: `-inf * 0` and `inf * 0` are `NaN` under IEEE-754, and `nan * 0` stays `NaN`.

### Scope: float32 only, measured

`bfloat16` is unaffected. Measured on ttsim 1.9.7, `Arch.BLACKHOLE`, `ttnn.triu(diagonal=1)` on a
32×32 tile, counting the 528 masked-out cells:

| input | dtype | zero | NaN |
|---|---|---|---|
| `-inf` | **float32** | 0 | **528** |
| `-inf` | bfloat16 | 528 | 0 |
| `-1e30` | float32 | 528 | 0 |
| `-1.0` | float32 | 528 | 0 |

The finite controls rule out the masking itself: it is specifically a non-finite input on the
`float32` path.

Worth stating plainly because the obvious motivating example does **not** apply — the CLIP
tutorial in `ttnn/tutorials/basic_python/` builds exactly this kind of additive `-inf` mask, but
it does so in `bfloat16`, so it is not affected. The case that bites is the same construction in
`float32`, which is the default an eager PyTorch port lands on.

### Why no test catches it

Every existing path draws finite inputs:

| coverage | range |
|---|---|
| `test_composite.py::test_unary_composite_{tril,triu}_ttnn` | `data_gen_with_range(..., -100, 100, ...)` |
| `tests/sweep_framework/sweeps/eltwise/unary/{tril,triu}` | `torch_random(low=-100, high=100)` |

With finite operands `multiply` and masking agree exactly, so the composition is correct
everywhere it is currently exercised.

### Fix

Select instead of multiplying:

```cpp
return ttnn::where(index_u, input_a, 0.0f, output_mem_config);
```

`where` with a scalar false-value is already the idiom in this file — `clamp` uses the same
`where(predicate, tensor, scalar, memory_config)` shape a hundred lines up — and
`TensorScalarVariant` takes the `float` directly.

The dispatch matters here, so I checked it rather than assuming. `where(Tensor, Tensor, float)`
resolves to the TTS overload in `ternary.cpp`, which goes to
`where_operation_with_scalar<WHERE_TTS>` and the SFPU kernel. It does **not** reach
`ternary_utils::where_impl`, which is itself built out of
`multiply(gtz(p), a) + multiply(lez(p), b)` and would reproduce this exact defect.

I verified the primitive on ttsim 1.9.7 (blackhole) rather than trusting that reading — SFPU
`where` with `-inf` on the taken branch, over 2048 elements:

| dest | taken branch | untaken branch | NaN in untaken |
|---|---|---|---|
| bfloat16 | `-inf` | `0.0` | **0 / 2048** |
| float32 | `-inf` | `0.0` | **0 / 2048** |

It selects; it does not arithmetically combine.

### Verification

I do not have hardware and could not build ttnn here, so the change itself is not executed
end-to-end. What is measured:

- The composition's arithmetic, against the registered goldens: `multiply(full(-inf), mask)`
  gives 15 NaN on a 5×5 with `diagonal=1`; `torch.triu` gives 0; `where(mask, input, 0)` matches
  the golden exactly. With finite inputs all three agree, which is the control.
- The SFPU `where` behaviour with `-inf`, on ttsim, in the table above.

What is not: that the patched file compiles, and that the device reproduces the torch result for
the full op. The added test in `test_composite.py` covers both on your CI.

### Scope

`tril` and `triu` only. `index_trilu` and the mask semantics are untouched, and finite inputs
produce bit-identical output — the masked lanes were already exactly `0.0` and now come from the
selection rather than from a multiply.

---

<sub>Written with AI assistance. I reviewed the change, ran the measurements above myself, and I
am responsible for the submission.</sub>

### CODE REVIEW

**Comment by EazyHood** [comment]
@jberkowitzTT could someone from metalium-developers-eltwise take a look at this one?

It replaces the mask multiply in tril/triu with ttnn::where, so a masked-out inf comes out
as 0 instead of NaN. Two files, and the test covers float32 and bfloat16.

Flagging that #52150 landed a couple of hours later with the same substitution, so two
reviewers don't end up duplicating the work.

**Comment by mateusznowakTT** [comment]
@EazyHood : What is performance impact of the change?

**Comment by EazyHood** [comment]
> @EazyHood : What is performance impact of the change?

Thanks for asking, and thanks again for the ttsim pointer on #51845 — that unblocked my
earlier measurements.

Starting with the part that costs something. For block-float inputs (`bfloat8_b`,
`bfloat4_b`) this change **does** move the op from the FPU to the SFPU. `multiply` forces
`fast_and_approximate_mode = true` whenever either operand is block-float
(`binary.cpp:861-862`, `is_block_fmt_inp ? true : fast_and_approximate_mode.value_or(false)`),
and `is_binary_sfpu_op` then returns false for `MUL` because the mask is bf16 so `a != b`
(`binary_ng_device_operation.cpp:30-31`). `WHERE_TTS` is unconditionally SFPU (same file,
`:65-67`). The tril sweep does cover `bfloat8_b`
(`sweeps/eltwise/unary/tril/tril.py:32`), so that path is live. I have no measured numbers
for it.

For bf16 and fp32 there is no change of unit. `multiply` defaults to
`fast_and_approximate_mode = false` (`binary.hpp:153`), and with that `is_binary_sfpu_op`
returns `!fast_and_approximate_mode` = true for `MUL`. SFPU before, SFPU after.

The second real cost is sharded. `where` is excluded from multi-tile processing —
`binary_ng_program_factory.cpp:1004`, `if (enable_multi_tile && !is_where_op)` — so with all
three tensors sharded, `num_tiles_per_cycle` goes 2 → 1 for bf16. Interleaved is 1 either
way, so it does not apply there.

Per tile, the where kernel uses three DST registers instead of two and does one `fill_tile`
the mul kernel does not (`eltwise_where_no_bcast.cpp:36-52`, indices `i*3`, `i*3+1`,
`i*3+2`). Minor, and possibly worth a follow-up: `fill_tile_init()` sits inside the per-tile
loop and looks hoistable.

What does not change: one device program per `tril` call either way, the same CBs (two inputs
plus output; the `0.0f` travels as a runtime arg), no typecast introduced (the mask is bf16,
so `typecast_predicate` in `ternary.cpp` does not fire), and the output dtype is preserved
(`binary_ng_device_operation.cpp:629`, `is_where_op ? dtype_b : dtype_a`).

What I cannot give you is wall clock. I do not have a Tenstorrent card. I looked for a
published FPU-vs-SFPU comparison for eltwise binaries in `tech_reports/` and
`perf_microbenchmark/` and did not find one, and I would rather say that than estimate. If
ttsim reports per-kernel SFPU cycle counts at this granularity, tell me and I will run both
variants through it; I have not confirmed that it does. `ttnn.tril` and `ttnn.triu` are
already registered in `tests/ttnn/profiling/ops_for_profiling.py` if a profile run on a
specific config would settle it.

For scale: `index_tril` is host-side (`functions.hpp:57-96`). Every call allocates the full
padded volume, fills it with a scalar triple loop, tilizes on host — a second full allocation
and copy — and then does a full-size H2D transfer. None of that changes here, and I would
expect it to dominate either kernel choice, though I have not measured the split.

On alternatives, I looked for a cheaper fix and did not find one. `0 * inf = NaN` is
IEEE-754, so no mask value works: 1e-38, 1e-30 and 1.0 all leave `-inf` unchanged, and 0.0 or
-0.0 give NaN. A different mask dtype does not help for the same reason. `fill_pad` covers
implicit tile padding, not a triangular region. And there is no FPU select to fall back on:
`WHERE_TTS`/`WHERE_TST` and `BITWISE_AND` all `TT_THROW` on the FPU path
(`binary_ng_utils.cpp:264-270`, `376-388`). `where` looks like the only op in the tree that
produces an exact zero without multiplying.

If the sharded or block-float regression matters, two options and I am happy to do either:
gate the `where` path to the cases that need it and keep `multiply` elsewhere, or drop the
`!is_where_op` exclusion in the program factory. There is also a larger follow-up — a
device-side fused `tril` that skips the host mask and the H2D entirely — but that belongs in
its own PR.

So: is the block-float FPU→SFPU move and the sharded 2 → 1 acceptable given the NaN fix, or
would you rather I gate the change? If you would rather see numbers first, name the bench and
config and I will run whatever you point me at.
